### ISSUE TITLE
Adjust PyMuPDF to MuPDF 1.20.0

### ISSUE THREAD

**Issue by mjg** [pymupdf/PyMuPDF#1715]
The first release candidate for MuPDF 1.20.0 has been released and brings upon us documented API changes as well as others.

In a first attempt, I tried to rebuild PyMuPDF 1.19.6 against MuPDF 1.20.0-rc1 with the obvious change
```
diff --git a/fitz/version.i b/fitz/version.i
index a002e4c..9f423c7 100644
--- a/fitz/version.i
+++ b/fitz/version.i
@@ -1,6 +1,6 @@
 %pythoncode %{
-VersionFitz = "1.19.0"
+VersionFitz = "1.20.0"
 VersionBind = "1.19.6"
 VersionDate = "2022-03-03 00:00:01"
 version = (VersionBind, VersionFitz, "20220303000001")
-%}
\ No newline at end of file
+%}
```
(i.e. without a PyMuPDF version change for now) and the somewhat less obvious change to get the swig bindings to build:
```
diff --git a/fitz/fitz.i b/fitz/fitz.i
index ce28a21..52a7eb7 100644
--- a/fitz/fitz.i
+++ b/fitz/fitz.i
@@ -8654,10 +8654,10 @@ struct Outline {
         }

         %pythoncode %{@property%}
-        int page()
+        PyObject *page()
         {
             fz_outline *ol = (fz_outline *) $self;
-            return ol->page;
+            return Py_BuildValue("ii", ol->page.chapter, ol->page.page);
         }

         %pythoncode %{@property%}
@@ -12850,4 +12850,4 @@ def __del__(self):
         self.__swig_destroy__(self)
         %}
     }
-};
\ No newline at end of file
+};
```
(`page` is a `struct fz_location` now with members `chapter` and `page` - what a choice ...).

Alas, the resulting build does not pass the test suite, apparantly because quite a bit has changed around document initialisation and outline iterators. This is over my head, and I entrust any further progress to upstream ... Let me know if I can help by running tests on Fedora infrastructure. Fow now, here ist the [build and test output](https://download.copr.fedorainfracloud.org/results/mjg/python-PyMuPDF/fedora-36-x86_64/04415623-python-PyMuPDF/builder-live.log.gz) from my [copr build](https://copr.fedorainfracloud.org/coprs/mjg/python-PyMuPDF/build/4415623/) against a test build of mupdf 1.20.0-rc1.

I will hold back releasing the Fedora package of mupdf 1.20.0 until (1.20.0 is released and) PyMuPDF is ready for it. (I have a fix for zathura-pdf-mupdf, the other dependent package).

**Comment by JorjMcKie** [comment]
Version 1.20.0 is in underway and will be published shortly.

**Comment by julian-smith-artifex-com** [comment]
FYI there's an early release candidate for PyMuPDF-1.20.0 available, which can be installed with:

    pip install pymupdf==1.20.0rc1

The usual wheels are available for Windows, MacOS and Linux. (Except there are no wheels for Linux/aarch64; they will be available in the next release candidate.)

This release does not have new PyMuPDF features, instead it's mainly concerned with working with the upcoming mupdf-1.20 release, along with some changes to how setup.py builds things.

**Comment by mjg** [comment]
> FYI there's an early release candidate for PyMuPDF-1.20.0 available, which can be installed with:
> 
> ```
> pip install pymupdf==1.20.0rc1
> ```
> 
> The usual wheels are available for Windows, MacOS and Linux. (Except there are no wheels for Linux/aarch64; they will be available in the next release candidate.)
> 
> This release does not have new PyMuPDF features, instead it's mainly concerned with working with the upcoming mupdf-1.20 release, along with some changes to how setup.py builds things.

Thanks, but we (Fedora) build and package from source, not from wheels. I don't see any related commits here. Is this built off a fork?

Possible changes to the build process (like you mentioned) are the very reason why I try to adjust packaging pro-actively.

**Comment by julian-smith-artifex-com** [comment]
Yes, it's currently being developed in a temporary repostiory: https://github.com/ArtifexSoftware/PyMuPDFTest/

Experimenting with github Actions for building wheels etc has entailed lots of test commits and rebasing that seemed best kept out of the main repository. All changes will be pushed to the main repository for the 1.20.0 release. And in future we'll use the main repository for all development as before.

**Comment by mjg** [comment]
Interesting. Having Artifex involved in PyMuPDF is certainly a good thing technically.

OTOH, the heavy rework on setup.py which makes it download mupdf sources during build makes me somewhat less enthusiastic, and it reminds me of the dreaded "bundle everything approach" which drives away every distro packager. I just hope that "proper builds" in isolated environments, without bundling, will still be possible and supported.

On a side note: "forking" in the GitHub sense or requesting branch access would have made these efforts more discovarable.

**Comment by julian-smith-artifex-com** [comment]
Thanks for the feedback. The bundling of mupdf was done to simplify our particular use case where we want wheel creation to be easy and reliable, but i don't want to preclude other ways of working, and i certainly don't want to make life difficult for distro packagers.

Setting PYMUPDF_SETUP_MUPDF_BUILD to an empty string will make PyMuPDF builds use the system mupdf. (Actually this wasn't quite working, but i've just pushed a fix to https://github.com/ArtifexSoftware/PyMuPDFTest.)

Does this address your concerns?

**Comment by mjg** [comment]
Thanks for being considerate about packagers ;-)

I had misread the new boilerplate of `setup.py`. With the latest push and setting `PYMUPDF_SETUP_MUPDF_BUILD` to an empty string, the package builds against mupdf-1.20.0-rc2 in a Fedora test environment (without internet access during the build process), and the test suite passes. Hooray!
[There are some failures on Fedora rawhide with "corrupted double-linked list" in the test suite which were there previously already.]

Fedora will test python 3.11 soon, so we will see how that goes. (The automateds test builds will be against the latest released versions, i particular PyMuPDF 1.19.6.)

**Comment by JorjMcKie** [comment]
I have already tested 3.11 with my local internal version 1.19.7 (which will not see the public) and Julian's pre-version 1.20.0. Both do work - as expected.
BTW for Python-heavy parts of PyMuPDF's code, I saw performance improvements very well in the range of the announced values of the [faster-cpython](https://github.com/faster-cpython/cpython) project: 20% or more. Among them are creating geometry objects (`Rect`, `Matrix`, etc.).
